File: Android/feature/feature-message/src/main/java/com/youkong/feature/message/viewmodel/ChatViewModel.kt
package com.youkong.feature.message.viewmodel

import android.util.Log
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.youkong.core.datastore.UserPreferences
import com.youkong.core.domain.manager.AppNotificationManager
import com.youkong.core.domain.manager.UnreadMessageManager
import com.youkong.core.domain.model.Message
import com.youkong.core.domain.model.MessageType
import com.youkong.core.domain.repository.MessageRepository
import com.youkong.core.websocket.WebSocketManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.datetime.Instant
import javax.inject.Inject

data class ChatUiState(
    val messages: List<Message> = emptyList(),
    val partnerName: String? = null,
    val currentUserId: String? = null,
    val isLoading: Boolean = true,
    val isSending: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class ChatViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val messageRepository: MessageRepository,
    private val userPreferences: UserPreferences,
    private val webSocketManager: WebSocketManager,
) : ViewModel() {

    private val conversationIdParam: String? = savedStateHandle["conversationId"]
    private val partnerIdParam: String? = savedStateHandle["partnerId"]

    private var conversationId: String? = null

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    private val messageIds = mutableSetOf<String>()
    private val messageMutex = Mutex() // 确保消息添加的原子性

    init {
        loadCurrentUser()
        initializeConversation()
        observeWebSocketMessages()
    }

    override fun onCleared() {
        super.onCleared()
        // 离开聊天页面时清除当前会话 ID
        AppNotificationManager.currentConversationId = null
        Log.d("ChatViewModel", "💬 [ChatViewModel] 清除当前会话ID")
    }

    private fun observeWebSocketMessages() {
        viewModelScope.launch {
            webSocketManager.newMessages.collect { newMessageData ->
                // 只处理当前会话的消息
                if (newMessageData.conversationId == conversationId) {
                    val message = convertToMessage(newMessageData.message)
                    addMessageIfNew(message)
                }
            }
        }
    }

    /**
     * 添加消息（如果不存在）
     * 使用 Mutex 确保原子性，避免并发添加导致的重复
     * 类似 iOS 的 addMessageIfNew 实现
     */
    private suspend fun addMessageIfNew(message: Message) {
        messageMutex.withLock {
            // 去重检查
            if (messageIds.contains(message.id)) {
                return
            }

            // 立即加入去重集合
            messageIds.add(message.id)

            // 更新消息列表
            _uiState.update {
                it.copy(messages = it.messages + message)
            }
        }
    }

    private fun convertToMessage(messageData: com.youkong.core.websocket.MessageData): Message {
        return Message(
            id = messageData.id,
            sender = com.youkong.core.domain.model.UserProfile(
                id = messageData.sender.id,
                nickname = messageData.sender.nickname,
                avatar = messageData.sender.avatar
            ),
            type = MessageType.valueOf(messageData.type),
            content = messageData.content,
            metadata = null,
            createdAt = Instant.parse(messageData.createdAt),
            isRead = messageData.isRead
        )
    }

    private fun initializeConversation() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            // 如果有 partnerId，先创建或获取会话
            if (partnerIdParam != null) {
                messageRepository.getOrCreateConversation(partnerIdParam)
                    .onSuccess { conversation ->
                        conversationId = conversation.id
                        _uiState.update { it.copy(partnerName = conversation.partner.nickname) }
                        loadMessages()
                    }
                    .onFailure { e ->
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                error = "创建会话失败: ${e.message}"
                            )
                        }
                    }
            } else if (conversationIdParam != null) {
                // 直接使用 conversationId
                conversationId = conversationIdParam
                loadMessages()
            } else {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        error = "缺少会话参数"
                    )
                }
            }
        }
    }

    private fun loadCurrentUser() {
        viewModelScope.launch {
            val userId = userPreferences.userId.first()
            _uiState.update { it.copy(currentUserId = userId) }
        }
    }

    private fun loadMessages() {
        val convId = conversationId
        if (convId == null) {
            _uiState.update {
                it.copy(
                    isLoading = false,
                    error = "会话ID无效"
                )
            }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            // ✅ 会话 ID 获取成功后，设置当前会话 ID 并清除未读计数
            Log.d("ChatViewModel", "💬 [ChatViewModel] 设置当前会话ID: $convId")
            AppNotificationManager.currentConversationId = convId

            Log.d("ChatViewModel", "💬 [ChatViewModel] 清除会话 $convId 的未读计数")
            UnreadMessageManager.clearUnread(convId)

            messageRepository.getMessages(convId)
                .onSuccess { messages ->
                    // 使用统一的添加方法逐个添加消息
                    messages.forEach { message ->
                        addMessageIfNew(message)
                    }
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                        )
                    }
                }
                .onFailure { e ->
                    // 如果会话不存在（可能被删除），尝试重新创建
                    if (e.message?.contains("会话不存在") == true || e.message?.contains("403") == true) {
                        if (partnerIdParam != null) {
                            // 如果有 partnerId，重新创建会话
                            recreateConversation()
                        } else {
                            _uiState.update {
                                it.copy(
                                    isLoading = false,
                                    error = "会话已失效，请从好友列表重新发起聊天"
                                )
                            }
                        }
                    } else {
                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                error = e.message ?: "加载失败"
                            )
                        }
                    }
                }
        }
    }

    private fun recreateConversation() {
        val partnerId = partnerIdParam ?: return

        viewModelScope.launch {
            messageRepository.getOrCreateConversation(partnerId)
                .onSuccess { conversation ->
                    conversationId = conversation.id
                    _uiState.update { it.copy(partnerName = conversation.partner.nickname) }
                    loadMessages()
                }
                .onFailure { e ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            error = "重新创建会话失败: ${e.message}"
                        )
                    }
                }
        }
    }

    fun sendMessage(content: String) {
        val convId = conversationId
        if (convId == null) {
            _uiState.update { it.copy(error = "会话ID无效") }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isSending = true) }

            messageRepository.sendMessage(
                conversationId = convId,
                type = MessageType.TEXT,
                content = content,
                metadata = null,
            ).onSuccess { message ->
                // 使用统一的添加方法，确保原子性
                addMessageIfNew(message)
                _uiState.update {
                    it.copy(
                        isSending = false,
                    )
                }
            }.onFailure { e ->
                _uiState.update {
                    it.copy(
                        isSending = false,
                        error = e.message ?: "发送失败"
                    )
                }
            }
        }
    }
}

File: iOS/YouKong/Presentation/Screens/Messages/ChatViewModel.swift
import Foundation
import Combine
import Factory

@MainActor
final class ChatViewModel: ObservableObject {
    @Published var messages: [Message] = []
    @Published var isLoading = false
    @Published var messageInput = ""
    @Published var isSending = false
    @Published private(set) var partnerName: String = "加载中..."
    @Published var respondingBookingId: String? = nil

    var canSendMessage: Bool {
        !messageInput.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !isSending
    }

    @Injected(\.messageRepository) private var repository
    @Injected(\.apiClient) private var apiClient

    private var partner: UserProfile?
    private var conversationId: String?
    private var messageIds = Set<String>()

    init(partner: UserProfile, conversationId: String?) {
        self.partner = partner
        self.partnerName = partner.nickname
        self.conversationId = conversationId
        setupWebSocket()
        updateCurrentConversation()
    }

    /// 从通知跳转时使用，只有 conversationId
    init(conversationId: String) {
        self.partner = nil
        self.conversationId = conversationId
        setupWebSocket()
        updateCurrentConversation()
    }

    deinit {
        // 离开聊天页面时清除当前会话 ID
        Task { @MainActor in
            NotificationManager.shared.currentConversationId = nil
        }
    }

    private func updateCurrentConversation() {
        if let conversationId = conversationId {
            NotificationManager.shared.currentConversationId = conversationId
        }
    }

    /// 判断消息是否来自对方
    func isFromPartner(_ message: Message) -> Bool {
        if let partner = partner {
            return message.sender.id == partner.id
        }
        // 如果没有 partner 信息，通过消息推断
        let currentUserId = AuthManager.shared.currentUser?.id ?? ""
        return message.sender.id != currentUserId
    }

    func loadMessages() async {
        if conversationId == nil {
            await createConversation()
        }

        guard let conversationId = conversationId else { return }

        // ✅ 会话 ID 获取成功后，更新当前会话 ID（防止收到消息时增加未读）
        updateCurrentConversation()

        // ✅ 清除未读计数
        print("💬 [ChatViewModel] 清除会话 \(conversationId) 的未读计数")
        UnreadMessageManager.shared.clearUnread(for: conversationId)

        isLoading = true
        defer { isLoading = false }

        do {
            let fetchedMessages = try await repository.getMessages(conversationId: conversationId)
            for message in fetchedMessages {
                addMessageIfNew(message)
            }

            // 如果没有 partner 信息，从消息中推断
            if partner == nil, let firstMessage = fetchedMessages.first {
                let currentUserId = AuthManager.shared.currentUser?.id ?? ""
                if firstMessage.sender.id != currentUserId {
                    partner = firstMessage.sender
                    partnerName = firstMessage.sender.nickname
                } else {
                    // 尝试从其他消息中找到对方
                    for message in fetchedMessages {
                        if message.sender.id != currentUserId {
                            partner = message.sender
                            partnerName = message.sender.nickname
                            break
                        }
                    }
                }
            }
        } catch {
            print("[ChatViewModel] Load messages error: \(error)")
        }
    }

    func sendMessage() async {
        let content = messageInput.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return }

        if conversationId == nil {
            await createConversation()
        }

        guard let conversationId = conversationId else {
            print("[ChatViewModel] Failed to create conversation")
            return
        }

        isSending = true
        messageInput = "" // 立即清空输入框

        do {
            let request = SendMessageRequest(
                type: .text,
                content: content,
                metadata: nil
            )
            let message = try await repository.sendMessage(
                conversationId: conversationId,
                request: request
            )
            addMessageIfNew(message)
        } catch {
            // 发送失败，恢复输入内容
            messageInput = content
            print("[ChatViewModel] Send message error: \(error)")
        }

        isSending = false
    }

    func respondToBooking(bookingId: String, action: String) async {
        respondingBookingId = bookingId
        defer { respondingBookingId = nil }

        do {
            let endpoint = APIEndpoint.respondToBooking(bookingId: bookingId, action: action)
            let _: BookingRespondResponse = try await apiClient.request(endpoint)
            await loadMessages()
        } catch {
            print("[ChatViewModel] Respond to booking error: \(error)")
        }
    }

    private func createConversation() async {
        guard let partnerId = partner?.id else {
            print("[ChatViewModel] Missing partner information")
            return
        }
        do {
            let conversation = try await repository.createConversation(partnerId: partnerId)
            self.conversationId = conversation.id
        } catch {
            print("[ChatViewModel] Create conversation error: \(error)")
        }
    }

    private func setupWebSocket() {
        WebSocketManager.shared.onMessage = { [weak self] convId, message in
            guard let self = self else { return }
            if convId == self.conversationId {
                self.addMessageIfNew(message)
            }
        }
    }

    private func addMessageIfNew(_ message: Message) {
        guard !messageIds.contains(message.id) else { return }
        messageIds.insert(message.id)

        if let index = messages.firstIndex(where: { $0.createdAt > message.createdAt }) {
            messages.insert(message, at: index)
        } else {
            messages.append(message)
        }
    }
}
